iio/iio.c

```diff
@@ -0,0 +1,838 @@
+
+	iio_interface = iio_get_interface(device, iio_interfaces);
+	if (!iio_interface)
+		return FAILURE;
+
+	return iio_rd_wr_attribute(&el_info, (char*)buf, len, iio_interface->iio, 1);
+}
+
+/**
+ * @brief Read channel attribute.
+ * @param *device - String containing device name.
+ * @param *channel - String containing channel name.
+ * @param *ch_out - Channel type input/output.
+ * @param *attr - String containing attribute name.
+ * @param *buf - Buffer where value is stored.
+ * @param len - Maximum length of value to be stored in buf.
+ * @return number of bytes read.
+ */
+static ssize_t iio_ch_read_attr(const char *device, const char *channel,
+				bool ch_out, const char *attr, char *buf, size_t len)
+{
+	struct element_info el_info;
+	struct iio_interface *iio_interface;
+
+	if (!iio_supported_dev(device))
+		return FAILURE;
+
+	el_info.device_name = device;
+	el_info.channel_name = channel;
+	el_info.attribute_name = attr;
+	el_info.ch_out = ch_out;
+
+	iio_interface = iio_get_interface(device, iio_interfaces);
+	if (!device)
+		return FAILURE;
+
+	return iio_rd_wr_attribute(&el_info, buf, len, iio_interface->iio, 0);
+}
+
+/**
+ * @brief Write channel attribute.
+ * @param *device - String containing device name.
+ * @param *channel - String containing channel name.
+ * @param *ch_out - Channel type input/output.
+ * @param *attr - String containing attribute name.
+ * @param *buf - Value to be written.
+ * @param len - Length of data in "buf" parameter.
+ * @return number of written bytes.
+ */
+static ssize_t iio_ch_write_attr(const char *device, const char *channel,
+				 bool ch_out, const char *attr, const char *buf, size_t len)
+{
+	struct element_info el_info;
+	struct iio_interface *iio_interface;
+
+	if (!iio_supported_dev(device))
+		return -ENODEV;
+
+	el_info.device_name = device;
+	el_info.channel_name = channel;
+	el_info.attribute_name = attr;
+	el_info.ch_out = ch_out;
+	iio_interface = iio_get_interface(device, iio_interfaces);
+
+	if (!iio_interface)
+		return -ENOENT;
+
+	return iio_rd_wr_attribute(&el_info, (char*)buf, len, iio_interface->iio, 1);
+}
+
+/**
+ * @brief Open device.
+ * @param *device - String containing device name.
+ * @param *sample_size - Sample size.
+ * @param *mask - Channels to be opened.
+ * @return SUCCESS, negative value in case of failure.
+ */
+static int32_t iio_open_dev(const char *device, size_t sample_size,
+			    uint32_t mask)
+{
+	struct iio_interface *dev;
+	uint32_t ch_mask;
+
+	if (!iio_supported_dev(device))
+		return -ENODEV;
+
+	dev = iio_get_interface(device, iio_interfaces);
+	ch_mask = 0xFFFFFFFF >> (32 - dev->iio->num_ch);
+
+	if (mask & ~ch_mask)
+		return -ENOENT;
+
+	dev->ch_mask = mask;
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Close device.
+ * @param *device - String containing device name.
+ * @return SUCCESS, negative value in case of failure.
+ */
+static int32_t iio_close_dev(const char *device)
+{
+	return iio_supported_dev(device) ? SUCCESS : FAILURE;
```

Comment:
shouldn't `iio_close_dev` also unset bits in the `dev->ch_mask` ?

Reply:
Set mask to 0.